jax/experimental/export/_shape_poly.py

```diff
@@ -495,56 +494,67 @@ def to_single_term(self) -> tuple[int, int, _DimMon] | None:
     assert mon is not None
     return (n1, n2, mon)
 
-  @classmethod
-  def add_coeff(cls, coeffs: dict[_DimMon, int], t: _DimMon, coeff: int):
+  @staticmethod
+  def add_coeff(coeffs: dict[_DimMon, int], t: _DimMon, coeff: int):
     """coeffs[t] += coeff, with squashing 0 coefficients."""
     if coeff == 0: return
-    n_coeff = coeffs.get(t, 0) + coeff
-    if n_coeff == 0:
-      del coeffs[t]
-    else:
-      coeffs[t] = n_coeff
+    coeffs[t] = coeffs.get(t, 0) + coeff
 
-  @classmethod
-  def normalize_coeffs(cls,
-                       coeffs: dict[_DimMon, int],
-                       scope: SymbolicScope) -> DimSize:
-    """Constructs a _DimExpr in normal form.
+  @staticmethod
+  def _normalize_term(t: _DimMon, t_k: int,
+                      scope: SymbolicScope) -> Sequence[tuple[_DimMon, int]]:
+    # If (t, t_k) is among the scope normalization rules, then return
+    # a list of updates to apply to the expression containing (t, t_k).
+    # Returns empty sequence if no normalizations are necessary.
+    updates = []
+    after, t_k_after = scope._normalization_rules.get(t, (None, 0))
+    if after is not None and t_k % t_k_after == 0:
+      # We have t*t_k_after -> after.
+      # We subtract `t*t_k` and add `after * (- (t_k // t_k_after))`.
```

Comment:
There is no `-` in the expression below. Is the negation implicit somehow?

Same for the other clause.

Reply:
Bad comments, fixed.